File: solutions/187.Repeated_DNA_Sequences/AC_bitwise_nlogn.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_bitwise_nlogn.cpp
*  Create Date: 2015-02-24 00:57:54
*  Descripton:  Bitwise, using STL set.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
private:
	int get_value(char ch) {
		switch (ch) {
			case 'A': return 0; break;
			case 'C': return 1; break;
			case 'G': return 2; break;
			case 'T': return 3; break;
		}
	}

public:
	vector<string> findRepeatedDnaSequences(string s) {
		set<long long> appeared, used;
		vector<string> ans;
		if (s.length() <= 10)
			return ans;

		long long hash = 0;
		// get first substring's hash
		for (int i = 0; i < 10; i++)
			hash = (hash << 2) | get_value(s[i]);
		appeared.insert(hash);

		// deal every substring
		for (int i = 10; i < s.length(); ++i) {
			hash = ((1LL<<20) - 1) & ((hash << 2) | get_value(s[i]));
			if (used.count(hash))
				continue;
			if (appeared.count(hash)) {
				ans.push_back(s.substr(i - 9, 10));
				used.insert(hash);
			} else
				appeared.insert(hash);
		}

		return ans;
	}
};

int main() {
	string ss;
	Solution s;
	while (cin >> ss) {
		vector<string> ans = s.findRepeatedDnaSequences(ss);
		for (auto &i : ans)
			cout << i << ", ";
		cout << endl;
	}
	return 0;
}

File: solutions/187.Repeated_DNA_Sequences/AC_bitwise_nlogn.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_bitwise_nlogn.py
# Create Date: 2015-02-25 11:34:46
# Usage:       AC_bitwise_nlogn.py 
# Descripton:  


class Solution:
    # @param s, a string
    # @return a list of strings
    def findRepeatedDnaSequences(self, s):
        value = {'A': 0, 'C': 1, 'G': 2, 'T': 3}
        appeared = set()
        used = set()
        ans = []

        if len(s) <= 10:
            return ans

        hash = 0
        for i in range(10):
            hash = (hash << 2) | value[s[i]]
        appeared.add(hash)

        for i in range(10, len(s)):
            hash = ((hash << 2) | value[s[i]]) & ((1<<20) - 1)
            if hash in used:
                continue
            if hash in appeared:
                ans.append(s[i - 9: i + 1])
                used.add(hash)
            else:
                appeared.add(hash)
        return ans

# debug
s = Solution()
print s.findRepeatedDnaSequences('AAAAAAAAAAAA')
